Cody: Add task view and non-stop cody recipe

abeatrix opened the pull request:
Close https://github.com/sourcegraph/sourcegraph/issues/51939

This PR introduces:
- A new feature flag called `cody.experimental.nonStop`
- If `cody.experimental.nonStop` is set to true, a task list view for fixup will be available under Cody's chat view
- A new recipe called non-stop cody
- A new command to run the non-stop cody recipe if the feature flag is on

NOTE: The primary goal of this OR is to add a task view for the future implementation for non-stop Cody. It does not include the ability to run multiple tasks at the moment. It only displays a list of tasks started by the non-stop recipe with their status.

## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

See loom for a detailed walkthrough of the changes introduced in this PR: https://www.loom.com/share/76ffab4b7e1c45148d300456452ec487

Update: based on Tim's feedback from Loom, I've removed the color from the task icons and moved the button to add a new task from the view title to the editor title:
<img width="1992" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/b428aaf1-1eff-47e8-8184-747f4ece89ed">


Update 2: loom on the changes included in the latest PR https://www.loom.com/share/202ef70b9f3a4cfe88ac0ab6f19e7181
![image](https://github.com/sourcegraph/sourcegraph/assets/68532117/82d7c0b9-6a89-4628-80aa-c794bf8a2138)

dominiccooney commented on client/cody-shared/src/editor/index.ts:
```diff
@@ -19,8 +19,15 @@ export interface InlineController {
+    refresh(): void
+}
+
 export interface Editor {
```
How much toil are we biting off here? Because we need to do text decorations and things later. I don't want to have to wrap a huge part of the VScode API through this abstraction layer. Can we assert a type somewhere so we don't have to keep elaborating this interface?

abeatrix replied:
I think this should be it? I was thinking the task itself handles all the range tracking as well as decorations, so we won't need to add those here, wdyt?

dominiccooney commented on client/cody-shared/src/chat/recipes/non-stop.ts:
```diff
@@ -0,0 +1,103 @@
+            return null
+        }
+
+        if (!context.editor.taskView) {
```
When does this happen in practice?

abeatrix replied:
on bad days 😢  but seriously it was just for type purpose 🤣

dominiccooney commented on client/cody-shared/src/chat/recipes/non-stop.ts:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        // Create a id using current data and use it as the key for response multiplexer
+        const taskID = Date.now().toString(36).replace(/\d+/g, '')
```
We are running in one address space here, just have a variable which you increment to assign IDs.

dominiccooney commented on client/cody-shared/src/chat/recipes/non-stop.ts:
```diff
@@ -0,0 +1,103 @@
+    If it doesn't make sense, you do not need to provide <selection-{taskID}>.
+
+    \`\`\`
+    {truncateTextStart}<selection-{taskID}>{selectedText}</selection-{taskID}>{truncateFollowingText}
```
Each LLM recipe turn gets its own BotResponseMultiplexer. These tags don't need to be unique. We can use the closure consuming the response to keep track of which response is for which task. I'm curious why we'd change the tag name.

abeatrix replied:
Oh right! I thought we will need a unique tag for each task but I completely forgot that this won't be necessary in our case where we will only run one task at a time and not concurrently 😅

dominiccooney commented on client/cody/src/editor/vscode-editor.ts:
```diff
@@ -9,9 +9,10 @@ import {
 
 export class VSCodeEditor implements Editor {
-    constructor(public controller: InlineController) {}
+    constructor(public controller: InlineController, public taskView: TaskViewProvider) {}
```
Let's use a parameter object here. These are only going to increase, and it will make unit testing easier if we can pass a bag of mocks at once; it will make the callsite easier to read; etc. etc.

abeatrix replied:
combined into a controllers object

dominiccooney commented on client/cody/src/services/TaskViewProvider/CodyTask.ts:
```diff
@@ -0,0 +1,183 @@
+    public selectionRange: vscode.Range | vscode.Selection = new vscode.Range(0, 0, 0, 0)
+    public state: CodyTaskState = CodyTaskState.queued
+    private readonly content: string
+    public readonly uri: string
```
For diff view we are going to have another related URI flying around, so let's give this a more specific name than just its type. It's the sourceDocumentUri or something like that.

dominiccooney commented on client/cody/src/services/TaskViewProvider/CodyTask.ts:
```diff
@@ -0,0 +1,183 @@
+    private replacementContent = ''
+    public errorMsg = ''
+    public readonly label: string
+    public selectionRange: vscode.Range | vscode.Selection = new vscode.Range(0, 0, 0, 0)
```
We will need a document change listener updating these ranges; they need to be that "tracked range" thing from my earlier prototype. This is fine for now but drop a TODO.

dominiccooney commented on client/cody/src/services/TaskViewProvider/CodyTask.ts:
```diff
@@ -0,0 +1,183 @@
+
+// TODO: Create CodeLens for each task
+export class CodyTask extends vscode.TreeItem {
+    public contextValue: string
```
What's this?

abeatrix replied:
It's a variable for the vscode tree item that is used to [contribute item specific actions in the tree](https://vshaxe.github.io/vscode-extern/vscode/TreeItem.html#contextValue) 😆  will add a comment to this

dominiccooney commented on client/cody/src/services/TaskViewProvider/CodyTask.ts:
```diff
@@ -0,0 +1,183 @@
+        super(input)
+        this.label = this.input
+        this.uri = selection.fileName
+        this.tooltip = `${this.label}-${this.uri}`
```
Let's have a DEBUG symbol somewhere and turn these on when it is set. So all of our debugging scree is evident because it will be guarded with if (cody.DEBUG) or something like that.

abeatrix replied:
updated to send debug logs to output channel

dominiccooney commented on client/cody/src/services/TaskViewProvider/CodyTask.ts:
```diff
@@ -0,0 +1,183 @@
+    constructor(public readonly id: string, private input: string, private selection: ActiveTextEditorSelection) {
+        super(input)
+        this.label = this.input
+        this.uri = selection.fileName
```
On the face of it this line isn't great... is it a filename, or a URI? Let's get away from that abstract editor API, I think it will make our lives worse and not better in the short term. In particular, we really need to sync changes back up with their document. So maybe we should hang onto the related editor, or the document, or the URI...

abeatrix replied:
Updated to use vscode.Uri

dominiccooney commented on client/cody/src/services/TaskViewProvider/CodyTask.ts:
```diff
@@ -0,0 +1,183 @@
+}
+
+// TODO: Create CodeLens for each task
+export class CodyTask extends vscode.TreeItem {
```
I think we should separate the task tracking object from the UX presentation of that object, instead of rolling them together. I made some comments on individual fields below... I won't go through and comment on each field, but I think we want to separate the object from its presentation and use MVVM or something like that. It keeps each concern separated and it makes it easy to, for example, write tests for the task processing part without having to set up a treeview and so on.

dominiccooney commented on client/cody/src/services/TaskViewProvider/CodyTask.ts:
```diff
@@ -0,0 +1,183 @@
+        this.set(CodyTaskState.done)
+    }
+    /**
+     * Get current selected lines from the comment thread.
```
I am a tiny bit regretting having fixup in comments at this point, because we now have a more difficult task for non-stop fixups. It would be great if we could just get the task management UI done, consuming mock data, without tying it to editors and so on yet.

A lot of this work will need a do-over to get the concurrent editing working and I think we need to divide and conquer that work. Time spent writing this might have been better spent working on tests for other pieces of it.

dominiccooney commented on client/cody/src/services/TaskViewProvider/CodyTask.ts:
```diff
@@ -0,0 +1,183 @@
+        return selection
+    }
+    /**
+     * Get Editor by URI
```
We should be able to get the editor for a given URI without making it the active editor. We should concentrate this diff on the task management UI. It's fine if it is consuming canned data, I'd just like to see that component really accurately producing @toolmantim 's design, with tests, and a nice separation between the presentation layer and the fixup execution, which can just be stubs for now.

abeatrix replied:
updated to get editor by URI

dominiccooney reviewed (commented):
Nice progress although I had in mind that this should really concentrate on the task management UI, with tests, and separation from the details of how the task execution, replacement, etc. happens. More comments inline.

abeatrix commented on client/cody-shared/src/chat/recipes/non-stop.ts:
```diff
@@ -0,0 +1,103 @@
+
+    public async getInteraction(humanChatInput: string, context: RecipeContext): Promise<Interaction | null> {
+        const selection = context.editor.getActiveTextEditorSelection() || context.editor.controller?.selection
+        if (!selection || !humanChatInput) {
```
Added input box

abeatrix commented:
Loom on the changes included in the latest PR https://www.loom.com/share/202ef70b9f3a4cfe88ac0ab6f19e7181
- display tasks by file path
- logs in output channel
- allow selection to be empty for making changes in empty files 
- allow human input to be empty for cases where instruction is included via comments in selected text
![image](https://github.com/sourcegraph/sourcegraph/assets/68532117/82d7c0b9-6a89-4628-80aa-c794bf8a2138)

dominiccooney reviewed (changes requested):
I think this still needs work. Where are the view/item/context contributions for the icons in the tree view in this design? I would love to see a screenshot with really high fidelity to this:

<img width="409" alt="Screenshot 2023-05-19 at 22 32 30" src="https://github.com/sourcegraph/sourcegraph/assets/55120/c113ecbe-0386-4a4c-843e-0e7d25adda98">

For example, that summary text "3 fixups, 1 running, 1 failed" is something you can derive from the data the view receives about the FUTs. So that's a concern that does fall within the view and should be handled here. But actually doing the churning of the tasks, that falls outside the widget and we don't want that in this patch.

I've sketched the reasons a bit, but to briefly elaborate, we need to divide and conquer this problem by making composable pieces, each one of high quality, which work well together. Because it's a complicated problem, and that helps us deal with the complexity. And later, we may want to assemble the pieces differently. For example, we should be able to churn fixup tasks even if there's no treeview. In that likely future scenario, we would regret having FUTs be tree view nodes.

abeatrix commented on client/cody-shared/src/chat/recipes/fixup.ts:
```diff
@@ -13,7 +13,7 @@ export class Fixup implements Recipe {
     public async getInteraction(humanChatInput: string, context: RecipeContext): Promise<Interaction | null> {
         // TODO: Prompt the user for additional direction.
-        const selection = context.editor.getActiveTextEditorSelection() || context.editor.controller?.selection
+        const selection = context.editor.getActiveTextEditorSelection() || context.editor.controllers?.inline.selection
```
This is due to the change where we now use a parameter object 
 instead of passing the views individually to the editor class. They are, however, not depending on each other. https://github.com/sourcegraph/sourcegraph/pull/51953#discussion_r1194710873

abeatrix commented:
> I think this still needs work. Where are the view/item/context contributions for the icons in the tree view in this design? I would love to see a screenshot with really high fidelity to this:
> 
> <img alt="Screenshot 2023-05-19 at 22 32 30" width="409" src="https://user-images.githubusercontent.com/55120/239547340-c113ecbe-0386-4a4c-843e-0e7d25adda98.png">
> For example, that summary text "3 fixups, 1 running, 1 failed" is something you can derive from the data the view receives about the FUTs. So that's a concern that does fall within the view and should be handled here. But actually doing the churning of the tasks, that falls outside the widget and we don't want that in this patch.
> 
> I've sketched the reasons a bit, but to briefly elaborate, we need to divide and conquer this problem by making composable pieces, each one of high quality, which work well together. Because it's a complicated problem, and that helps us deal with the complexity. And later, we may want to assemble the pieces differently. For example, we should be able to churn fixup tasks even if there's no treeview. In that likely future scenario, we would regret having FUTs be tree view nodes.

Thanks for clarifying! This makes sense to me! I do have a question regarding the image you are sharing: where can I find it lol I looked through the figma docs but I can't seem to find it lol  https://www.figma.com/file/9112BsKsJc1BpO2j8XYwLL/%F0%9F%A4%96-Cody-VS-Code-%5BMain%5D?type=design&node-id=2873-114601&t=toG1izmFRrxkjsiv-0

abeatrix commented:
@dominiccooney I've updated the Tree View based on Tim's designs and your feedback. Added summary text, created a new Controller that acts as the Factory for tasks, so task view now only handles turning tasks sent from the Controller into Tree Item. 

Since we don't have the diff view implemented for non-stop cody, I have replaced the accept-all button with the trash-can icon for now.

Also, just so you know, unlike in Tim's designs, directories do not have file extension icons, even if we are putting file names on them. 

Here are some screenshots from the latest commit:
<img width="368" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/aec049d4-88f4-4beb-8925-1bdc563014d1">
<img width="376" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/1735aede-edd1-43bc-b531-f3b35db1ad8d">
<img width="377" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/c071bcb3-2871-466b-9e2c-f6f9267e1ddd">
<img width="371" alt="image" src="https://github.com/sourcegraph/sourcegraph/assets/68532117/a27b9a92-701e-4d15-a86a-5ec8b53eac37">

Loom: https://www.loom.com/share/3686fcaa965e413b8de2f49554a46fdd

abeatrix commented:
Because the main branch has changed so much, moving this into a new PR instead https://github.com/sourcegraph/sourcegraph/pull/52282
